### DISCUSSION TITLE
Question: Should the client attempt reconnection if the graphql server goes down?

### DISCUSSION THREAD

**Discussion by brettjashford** [enisdenjo/graphql-sse#24] [Q&A] [answered]
should the client attempt to reconnect if the graphql server goes down (current behavior with `graphql-ws`)?

if i kill my graphql server, the stream closes without any attempt to reconnect. it seems that we would want [this `catch` block](https://github.com/enisdenjo/graphql-sse/blob/master/src/client.ts#L451) to be hit so the retry logic kicks in, but the async iterator completes and the loop exits. seems like a potential bug, but maybe expected behavior? 

i'm running in distinct connections mode, but seeing the same behavior in single connection mode.

- thanks 🙏

**Comment by enisdenjo** [comment]
So what's happening, when the server goes away, is instead of reaching [this `catch` block](https://github.com/enisdenjo/graphql-sse/blob/16f6a6c5ec77f63d19afda1c847e965a12513fc7/src/client.ts#L442-L451), the [`getResults` loop](https://github.com/enisdenjo/graphql-sse/blob/16f6a6c5ec77f63d19afda1c847e965a12513fc7/src/client.ts#L435-L438) exits and we reach [this `return` instead](https://github.com/enisdenjo/graphql-sse/blob/16f6a6c5ec77f63d19afda1c847e965a12513fc7/src/client.ts#L440)? If so, this feels like a bug - could you additionally hook a `console.log` in the [`onMessage` option](https://github.com/enisdenjo/graphql-sse/blob/master/docs/interfaces/ClientOptions.md#onmessage) and provide me with its logs?

**Reply by brettjashford** [reply]
when logging `onMessage`, i only get `next` event(s) (if received through the subscription) ex. 
```
{
    "event": "next",
    "data": {
        "data": ...
        }
    }
}
```
> instead of reaching [this catch block](https://github.com/enisdenjo/graphql-sse/blob/16f6a6c5ec77f63d19afda1c847e965a12513fc7/src/client.ts#L442-L451), the [getResults loop](https://github.com/enisdenjo/graphql-sse/blob/16f6a6c5ec77f63d19afda1c847e965a12513fc7/src/client.ts#L435-L438) exits and we reach [this return instead](https://github.com/enisdenjo/graphql-sse/blob/16f6a6c5ec77f63d19afda1c847e965a12513fc7/src/client.ts#L440)

i assumed that's what was happening, but after tracing through i see it hits [this `finally`](https://github.com/enisdenjo/graphql-sse/blob/master/src/client.ts#L730) where it calls the `proceed` callbacks and just kinda quits, never hits [this catch block](https://github.com/enisdenjo/graphql-sse/blob/16f6a6c5ec77f63d19afda1c847e965a12513fc7/src/client.ts#L442-L451) or [this return](https://github.com/enisdenjo/graphql-sse/blob/16f6a6c5ec77f63d19afda1c847e965a12513fc7/src/client.ts#L440). i'm kinda confused tbh, still getting used to async iterators

**Reply by enisdenjo** [reply]
Could it be this #22? But that one never hits the `finally` block. 🤔

**Reply by brettjashford** [reply]
possibly, i'll keep an eye on https://github.com/enisdenjo/graphql-sse/pull/23 and retest once it's solved

**Reply by enisdenjo** [reply]
Hey @brettjashford, #23 seems to happen only in Node environments and the issue might've not even be caused by `graphql-sse` itself.

I did a few tests in the browser myself, here's what I did:
1. Start a server with `graphql-sse`
2. Run GraphiQL using [this gist](https://gist.github.com/enisdenjo/d7bc1a013433502349d2763c3d2f2b79)
3. Subscribe to any subscription
4. Wait for 1st event emission
5. Kill server
6. The browser reports `POST http://localhost:4000/graphql/stream net::ERR_INCOMPLETE_CHUNKED_ENCODING 200 (OK)` immediately in the console

Can you confirm whether is the behaviour you're experiencing too?

**Comment by enisdenjo** [comment, marked as answer]
Should be fixed in latest versions with https://github.com/enisdenjo/graphql-sse/issues/28, https://github.com/enisdenjo/graphql-sse/issues/27 and https://github.com/enisdenjo/graphql-sse/commit/607b4684b99284108ba4cd4d9c79f0a56bd16fbd.
